Plugin Library build fixes

gillkyle opened the pull request:
This is a continuation of #3456 

A new dependency release and a few files (`www/gatsby_node.js` and `yarn.lock`) caused issues with the Travis CI build. I forked the most recent branch from jastack and resolved those issues and removed the `package-lock.json` file. 

There are still some things from the original thread (#3456 <-- that PR # is conveniently easy to type) that need to be resolved like some UI issues - mobile support, etc. - that I thought could be included later or summed up here and then worked on. I could also make a project board with all the remaining notes that need to be addressed if that would be helpful.

If there are other changes I should add before this moves forward I'd be happy to keep working on them!

KyleAMathews commented:
Cool! Thanks!

A project board would be great! We should definitely start using those more.

Assuming Travis + Netlify are happy, let's dark deploy this and then start working on remaining items.

KyleAMathews commented on packages/gatsby-source-npm-package-search/gatsby-node.js:
```diff
@@ -0,0 +1,81 @@
+"use strict";
```
This file shouldn't be checked in — copy the the .gitignore file from plop-templates/packages/gitignore.hbs

gillkyle replied:
The contents of the .gitignore in both places (plop-templates/packages/gitignore.hbs and packages/gatsby-source-npm-package-search/.gitignore) are the same. Is the handlebars file special in some way? Or would it be enough to just add `gatsby-node.js` to the .gitignore that's already there?

KyleAMathews replied:
you probably just need to run `git rm --cached gatsby-node.js` to remove it. If a file is checked in, even if it's now ignored by a .gitignore, it'll stay in Git until explicitly removed.

KyleAMathews commented on packages/gatsby-source-npm-package-search/package.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "gatsby-source-npm-package-search",
+  "version": "1.0.1",
+  "description": "Stub description for gatsby-source-npm-package-search",
```
Fix description

KyleAMathews commented on packages/gatsby-source-npm-package-search/src/gatsby-node.js:
```diff
@@ -0,0 +1,79 @@
+    hitsPerPage: 1000,
+  })
+
+  // Currently the gatsby-transformer-remark plugin
```
@gillkyle make sure to add this as a TODO to fix this in gatsby-remark-images or wherever the error is. Though I think @calcsam might have fixed this recently?

KyleAMathews commented on packages/gatsby-source-npm-package-search/src/gatsby-node.js:
```diff
@@ -0,0 +1,79 @@
+      return
+    }
+
+    const parentId = `plugin ${hit.objectID}`
```
Use the new createNodeId for this https://github.com/gatsbyjs/gatsby/pull/3683

KyleAMathews commented on packages/gatsby-source-npm-package-search/src/gatsby-node.js:
```diff
@@ -0,0 +1,79 @@
+
+    const parentId = `plugin ${hit.objectID}`
+    const readmeNode = {
+      id: `readme ${hit.objectID}`,
```
also here

KyleAMathews commented on packages/gatsby-source-npm-package-search/src/gatsby-node.js:
```diff
@@ -0,0 +1,79 @@
+const algoliasearch = require(`algoliasearch`)
+const crypto = require(`crypto`)
+
+var client = algoliasearch(`OFCNCOG2CU`, `f54e21fa3a2a0160595bb058179bfb1e`)
```
Use Const

KyleAMathews replied:
Get the values here from plugin options

KyleAMathews replied:
Or... maybe not actually — it'd be nice if this was just plug-n-play. Let's leave them hard-coded for now so people can just install the plugin and it works.

KyleAMathews commented on packages/gatsby-source-npm/gatsby-node.js:
```diff
@@ -0,0 +1,76 @@
+const algoliasearch = require(`algoliasearch`)
```
Needs gitignored and removed as well

gillkyle commented:
Awesome, I'll go through the original thread and pull out everything I find into a project board. I'm meeting up with @shannonbux tomorrow to talk about mobile layout and then I can get to work on some of the remaining issues.

KyleAMathews commented on packages/gatsby-source-npm/package.json:
```diff
@@ -0,0 +1,13 @@
+{
```
This looks like the old version of the plugin which should be removed.

KyleAMathews commented on www/gatsby-node.js:
```diff
@@ -6,6 +6,11 @@ const fs = require(`fs-extra`)
 
+const localPackages = `../packages`
+const localPackagesArr = []
+fs.readdirSync(localPackages).forEach(file => {
```
This wasn't finished — the idea here is for packages in this repo to use that README instead of what's grabbed from Algolia. The reason is that the Algolia README often doesn't have good links to images.

So take this `localPackagesArr` and use it to filter out the result of the query below getting all NPM package names.

gillkyle replied:
I still don't understand exactly what needs to change on this, do you mean `localPackagesArr` would give the readme data to the `template-docs-packages.js` file so that we don't have to rely on Algolia for the readmes? From what I understand that's what it's doing now. So the data displayed would be from the local folders rather than Algolia's search of npm that struggles with images?

KyleAMathews replied:
Yeah — Algolia (via NPM) has a copy of the READMEs for packages in this repo. We'd prefer to use those as they any referenced images.

gillkyle replied:
I came back to this this evening to look into finishing it up, sorry for the confusion on this, I want to make sure I can wrap my head around the actual issue though.

As of now `www/gatsby-node` is creating all the pages for the packages library based off of its own query of `allNpmPackage` (which is just getting all the names to create the pages, when I run that query in GraphiQL I don't see html returned for all the packages, just a handful of them). So to fill in the content on each page the `template-docs-packages.js` has it's own query that's filling the content on the screen from the `markdownRemark` type. Apparently there are packages that have images in their readme and they're causing errors? Do we know what any of those packages are?

KyleAMathews replied:
You should create two template components. One for local packages and the other for remote. They'd share the same components for rendering data but the local package would query the local README while remote packages would query the NPM readme. Make sense?

So when you're creating pages, filter out of the NPM packages anything in `localPackagesArr` and then create pages for `localPackagesArr` w/ the local package template.

KyleAMathews commented on www/src/assets/down-arrow-white.svg:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
We should use react-icons for this https://github.com/gorangajic/react-icons

KyleAMathews commented on www/src/css/searchbox-style.css:
```diff
@@ -0,0 +1,935 @@
+.ais-InstantSearch__root {
```
This all should be done with Glamor using the `css` prop.

KyleAMathews replied:
Actually... not even sure what this is

gillkyle replied:
This file looks like it was scss or less transpiled to css. Is it worth refactoring over into Glamor or leave it be?

KyleAMathews replied:
First question is double-check it's even being used. It's not referenced anywhere in this PR

gillkyle replied:
I'm not sure why it's not referenced but as far as I know this file is only on this branch. It's adding a lot of styles used here: https://yarnpkg.com/en/ on things like the search bar. Some of the styles aren't really important so I'm sorting through it and I'll see if some could be glamor-ified.

KyleAMathews replied:
They're not imported anywhere so they're not used so just delete the file.

gillkyle replied:
It looks like it was imported to the /layouts/index.js file. I took out references to classes that weren't styling anything so it's a little shorter, but deleting the file left things like the search bar in their natural-plain-browser state.

gillkyle replied:
I cut out like ~300ish lines of styles here, there are like 500 lines that are applied to components that are used. I think the easiest way to add them in as component props would be convert the css back to scss and then use those styles adding them as props to the react instant search components we're using. Should I go through and do that? Is there a need to style them that way over the stylesheet?

KyleAMathews replied:
no it's ok for now I guess. Feels like an absurd amount of CSS but whatever. We can refactor it later.

KyleAMathews commented on www/src/html.js:
```diff
@@ -71,11 +71,11 @@ export default class HTML extends React.Component {
             rel="mask-icon"
             href={`/safari-pinned-tab.svg`}
-            color={colors.gatsby}
+            color="#5bbad5"
```
Why was this color changed?

KyleAMathews commented on www/src/html.js:
```diff
@@ -71,11 +71,11 @@ export default class HTML extends React.Component {
           />
-          <meta name="msapplication-config" content={`/browserconfig.xml`} />
-          <script src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js" />
+
```
This shouldn't have been removed

gillkyle commented:
Awesome, I'll get on these!

fk commented:
Excited to see this moving forward! Thank you @gillkyle!

KyleAMathews commented:
Deploy preview for *using-glamor* failed.

Built with commit 4479e4cbe6c26e4a354f78773044248cb0db0071

https://app.netlify.com/sites/using-glamor/deploys/5a7d360192226a45cf0ab5e5

KyleAMathews commented:
Deploy preview for *using-glamor* failed.

Built with commit dafdbb4231d646c6a76e4786906a0abfbfbdce1e

https://app.netlify.com/sites/using-glamor/deploys/5a7d3674a6188f1fdc73030c

KyleAMathews commented:
Deploy preview for *using-glamor* failed.

Built with commit 6128a864c44425db20b416cd92390c02b31e6cab

https://app.netlify.com/sites/using-glamor/deploys/5a7d37037b6ee876d7ded5b9

gillkyle commented:
One issue I noticed was on packages that aren't located here: https://github.com/gatsbyjs/gatsby/tree/master/packages and don't have any data in the repository part of the query either, so there's no github link for them, throwing errors. Not sure why no information is returned in the repository part of the query:
```
query npmPackages {
  allNpmPackage {
    edges {
      node {
        id
        title
       	repository {
          url
          project
          user
        }
      }
    }
  }
}
```

calcsam commented:
@gillkyle could be a number of issues -- npm plugins aren't required to be backed by a Github repo so that's probably why the data isn't populated. If there aren't a ton of them feel free to omit creating pages

gillkyle commented:
Sounds good @calcsam, and it's the pageQuery for the `template-docs-packages.js` file @KyleAMathews. After merging in the upstream changes there's a little more data returned in that query for npmPackages but there are still some packages that have things missing so I'm trying to prevent undefined props being accessed as the pages component renders. Should packages like gatsby-typegen packages even be included in the search results when they don't have a docs page (at least in the master branch they don't)? Is that what you meant to filter our the results with the npm package names from the gatsby repo?

gillkyle commented:
This is what it's displaying now for the typegen packages: 
![screen shot 2018-02-14 at 1 42 13 pm](https://user-images.githubusercontent.com/21114044/36227024-f0d7583e-118c-11e8-8816-340b94205b75.png)

gillkyle commented:
![mobile_plugin_library](https://user-images.githubusercontent.com/21114044/36286681-01a4685a-126e-11e8-80f6-d0ebfa77b084.gif)
I did some tweaking to just stack the search bar over the readme file for mobile views, it seemed like the simplest implementation to get a mobile view and when you select a search result it takes you down to the read me. I was thinking maybe it'd be nice to have an up arrow to take you back up to the search bar since if you scroll down in the results list before tapping a result you'll hit the results list while scrolling back up and have to scroll it all the way to the top. Maybe that would be overkill but that was one idea. 

I also left the packages that are pulled in by algolia and aren't on the Gatsby docs for now (they saw readme not found like in the screenshot above).

Not sure what else needs to be taken care of though I made a project board on my fork so I could note some little things I thought might to be taken care of: https://github.com/gillkyle/gatsby/projects/1

KyleAMathews commented:
`gatsby-typegen-filesystem` is supposed to be deprecated but wasn't. I just deprecated it so I assume it'll go away now (we're filtering out deprecated packages I assume?).

gillkyle commented:
I actually don't think we are filtering out deprecated packages. I don't remember seeing it anywhere, I'll go change it to make sure.

KyleAMathews commented:
Instead of scrolling down — I'd just hide the search altogether. Then the user can just click the back button to return to the search.

gillkyle commented:
Using the browser or Android back button then? So almost like the search bar is it's own page?

KyleAMathews commented:
The search bar is it's own page :-) We just keep it on the screen when navigating to the package readmes.

KyleAMathews commented:
>Not sure what else needs to be taken care of (@calcsam / @KyleAMathews ) though I made a project board on my fork so I could note some little things I thought might to be taken care of

I think the most compelling thing there is moving the scss into glamor. It'll make it way easier for people in the future to jump in and make changes.

That can wait for a follow-up PR though.

fk commented:
@gillkyle This is a little off-topic but I just saw your project board (👍👍) and this todo:

![image](https://user-images.githubusercontent.com/21834/36290043-b988b4f8-12c3-11e8-9907-3bd03f7d1688.png)

Wanted to quickly let you know that I think I "fixed" that today in https://github.com/gatsbyjs/gatsby/pull/4059/commits/9c692b58d228cdcc2379380d0770e3f0d19ffe26…and now that I actually read the full error message saw that I should have used `undefined` instead of an empty string. 😁 😓 Anyways…

Also…we're close!! 💪

gillkyle commented:
Thanks @fk! Helping me move to-dos without me actually doing anything!

![mobile_plugin_library_2](https://user-images.githubusercontent.com/21114044/36346741-9991ab16-1401-11e8-8a59-481bb33dfe1f.gif)
Here's a different solution to the mobile layout. This starts getting a little confusing in the layout file with how it's hiding the sidebar and displaying the contents, so I tried my best to write it so it'd make sense. Maybe I need to add some comments?

I'll start refactoring the css now if that's the next best step.

fk commented:
Just had the chance to check this out locally—looking really good already!
Unfortunately I can not remember the previous version of the mobile UI (if there was any?), but what I just saw feels quite complete IMO.

For anyone who wants to follow along, I had to do a `yarn run bootstrap` in the repo root and `cd www && yarn add date-fns react-instantsearch`.

gillkyle commented:
@KyleAMathews @calcsam I made quite a few changes to how the templates created pages so that they'd be split up more readably between remote and local packages (ie packages from algolia, and packages in the filesystem). 

The css was pared down and included in CSS props where possible now too.

Let me know if anything else needs to change 🙂

calcsam commented:
@gillkyle awesome! could you fix the build errors! looks like we need to add date-fns to package.json, plus possible some other errors to fix?

https://app.netlify.com/sites/gatsbyjs/deploys/5a949cc0a6188f10940f0bbd

gillkyle commented:
@calcsam I added two missing dependencies but looks like building the static html is still failing. I'm trying to see what the deal is there. 

This error: `Error: Could not find function loadNodeContent for plugin gatsby-source-npm-pa  ckage-search` is one I don't see locally so I'm not sure what issue is causing it. Looking into both...

KyleAMathews commented on packages/gatsby-source-npm-package-search/src/gatsby-node.js:
```diff
@@ -0,0 +1,83 @@
+    hitsPerPage: 1000,
+  })
+
+  // TODO: Currently the gatsby-transformer-remark plugin
```
This needs fixed

gillkyle replied:
@KyleAMathews I pulled this out and found the problem packages, in terms of fixing is something like a little regular expression that replaces the badges and images with empty strings in the readme's content (so the markdown can still function) good?

KyleAMathews commented on packages/gatsby-source-npm-package-search/src/gatsby-node.js:
```diff
@@ -0,0 +1,83 @@
+      internal: {
+        type: `NPMPackageReadme`,
+        mediaType: `text/markdown`,
+        content: hit.readme,
```
To fix the `loadNodeContent` error — check that hit.readme is always defined as a string — so check if it's not defined and if it isn't, set it as any empty string.

KyleAMathews commented:
It's alive!!! https://deploy-preview-3906--gatsbyjs.netlify.com/packages/ :-D

gillkyle commented:
A couple of those little badges look a little wonky after I took out the snippet that was skipping the readmes with images: 

- [firestore](https://deploy-preview-3906--gatsbyjs.netlify.com/packages/gatsby-source-firestore)
- [trello](https://deploy-preview-3906--gatsbyjs.netlify.com/packages/gatsby-source-trello)

But they seem to work?..

KyleAMathews commented:
Yeah, we should be skipping any READMEs.

The badges seemed affected by our default underline CSS for links — we could override that for these pages.

calcsam commented:
soooo beautifulllll

On Thu, Mar 1, 2018 at 5:01 PM, Kyle Mathews <notifications@github.com>
wrote:

> Yeah, we should be skipping any READMEs.
>
> The badges seemed affected by our default underline CSS for links — we
> could override that for these pages.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gatsbyjs/gatsby/pull/3906#issuecomment-369784894>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEL3W8xiMCSmU4Pi8Dj8sHdiMi1pKBVIks5taJn4gaJpZM4R9oo8>
> .
>

shannonbux commented:
wow @gillkyle it looks awesomeeeeeeeeee. The search runs so fast I almost
thought it wasn't working lol! So speedy! Very excited!!!

On Thu, Mar 1, 2018 at 6:32 PM, Sam Bhagwat <notifications@github.com>
wrote:

> soooo beautifulllll
>
> On Thu, Mar 1, 2018 at 5:01 PM, Kyle Mathews <notifications@github.com>
> wrote:
>
> > Yeah, we should be skipping any READMEs.
> >
> > The badges seemed affected by our default underline CSS for links — we
> > could override that for these pages.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/gatsbyjs/gatsby/pull/3906#issuecomment-369784894>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AEL3W8xiMCSmU4Pi8Dj8sHdiMi1pKBVIks5taJn4gaJpZM4R9oo8>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gatsbyjs/gatsby/pull/3906#issuecomment-369790458>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Ae9o2uy9h8YJalSPtjNxruZQ4NP9Y4PFks5taKE8gaJpZM4R9oo8>
> .
>

calcsam commented:
It doesn't seem to load all the examples? Eg https://deploy-preview-3906--gatsbyjs.netlify.com/packages?=source doesn't show 51 results, it shows only 20 or so

gillkyle commented:
@calcsam yeah I had noticed that. I switched out the widget we were using for one that allows infinite scroll via a load more button (so it's basically paginated for every 20 results now). However,  changing that added some issues to how the selected item in the search list works that I'm trying to sort out, fixing one thing just adds another issue 😅

calcsam commented:
Haha well almost there! The pagination looks nice. Let us know when the
selection issue is fixed and then we can merge in!

On Sat, Mar 3, 2018 at 4:17 PM, gillkyle <notifications@github.com> wrote:

> @calcsam <https://github.com/calcsam> yeah I had noticed that. I switched
> out the widget we were using for one that allows infinite scroll via a load
> more button (so it's basically paginated for every 20 results now).
> However, changing that added some issues to how the selected item in the
> search list works that I'm trying to sort out, fixing one thing just adds
> another issue 😅
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gatsbyjs/gatsby/pull/3906#issuecomment-370190783>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEL3W3hYJEhM2p8zxA76rUzQf1ZunNTsks5tazKsgaJpZM4R9oo8>
> .
>

gillkyle commented:
@calcsam @KyleAMathews I fixed the highlighting issue, I refactored how the pages for each package were created too because a few pages (I think it was 4) were getting skipped over and instead added in some default not found data. 

I think it's about there! 🤞

calcsam commented:
Cool! Merging.
